# Pandas Intro and Basic File Management
[Basic OS Management](#BasicOSManagement)
[Read Data](#ReadData)

<a id="BasicOSManagement"></a>
## 1. Basic OS Management in Notebook
### Typically ALL the imported packages should be together in a single cell at the top of your notebook

In [4]:
import os

## What is the current working directory (folder) of this notebook and what’s actually in it (a list of files and folders)?
### This will show where you are currently working within your computer's folders. Make sure you know where this folder is. look at the way the path is written in the output that is the form that you need to use. It is different for Windows and Macs

In [5]:
os.getcwd()

'C:\\Users\\mdb025\\Documents\\Courses\\Analytics Course Materials\\Machine Learning\\ANOP 330\\ANOP330_F2021\\ANOP330F21_Notebooks'

In [7]:
os.listdir()

['.ipynb_checkpoints',
 'Assignment_HealthInsuranceCost_EDA_Reg_Soln.pdf',
 'CO2_DataViz.ipynb',
 'Data',
 'F_MNIST_Data.png',
 'HealthInsuranceCost_EDA_Reg_Soln.ipynb',
 'HW1_PandasIntro_soln.ipynb',
 'HW1_PandasIntro_soln_PDF.pdf',
 'HW1_PandasIntro_std.ipynb',
 'InClassNotes_9_9.ipynb',
 'NBA_Salary_Regressoin_SOLN_v2.pdf',
 'NeuralNetworks_MLP_FashionExp_1.ipynb',
 'Overview Pandas and Partition v1.ipynb',
 'Overview Pandas and Partition v2.ipynb',
 'PandasIntro_Inprogress_Notebook_8_27.ipynb',
 'plot_mnist_filters.ipynb',
 'ProfBaileyNotesPandasIntro.ipynb',
 'Python_regression_results.ipynb',
 'RegressionPractice',
 'Submit.csv',
 'TitanicLogitClassifierAssignment_Soln_PDF.pdf',
 'TitanicLogReg_soln-Copy1.ipynb',
 'TitanicLogReg_soln.ipynb',
 'TitanicRulePredAssignment_soln.ipynb',
 'TitanicRulePredAssignment_Soln_PDF.pdf',
 'Untitled.ipynb',
 'VehiclesEDA_Assignment_Soln.ipynb',
 'VehiclesEDA_Assignment_Soln_PDF.pdf']

## I want to move to a new folder (change directory)! Provide a string with the path.
(tip: If you copy and paste from office or the web, the quote characters may be copied incorrectly (“ vs. ") and that wlll throw an error. Just retype the quotes to fix.)

In [ ]:
os.chdir("C:\\Users\\mdb025\\Desktop")
#can view what's in the folder if wanted
#os.listdir()

## You could also create a path object, check it exists, and move there
### Need to import the function Path from the pathlib package (if you don't want to import an entire package and ALL it's functions, you can just import the ones you need using `from [package name] import [function(s) names]`.

In [16]:
from pathlib import Path
p = Path('C:\\Users\\mdb025\\Documents\\Courses\\Analytics Course Materials\\Machine Learning')
p.exists()

True

In [17]:
#move working directory to this path
os.chdir(p)

<a id="ReadData"></a>
## 1. Read in Data
### Now move to to your original directory (folder) where your Data folder exists and holds the data file `movie.csv`

In [19]:
os.chdir('C:\\Users\\mdb025\\Documents\\Courses\\Analytics Course Materials\\Machine Learning\\ANOP 330\\ANOP330_F2021\\ANOP330F21_Notebooks')

### If the file is in the working directory, then you can just open it by name, otherwise, you may need to add part of the (or the full) path to the file. It's not a bad idea to check the file exists before you try to open it.

In [20]:
os.path.isfile("Data\\movie.csv")

True

### We are going to import using common aliases for our bread and butter packages `pandas` and `numpy`

In [21]:
import pandas as pd
import numpy as np

### Read the data inta a pandas DataFrame (from a downloaded file or even a file from the web). Common convention is to add `df_` as a prefix or `_df` as a suffix to the name for clarity. You'll see both conventions when you look at sample code. Different data file types require different read functions (or options). `read_csv`, `read_excel`, and `read_json` or the most common. There are alternatives for database files also. We'll mostly stick with comma-separated values (CSV) files.
### Whenever we see a new function, do not assume that all you see is all there is, there are MANY optional inputs that you should be exploring! Click and review!
[pandas.read_csv](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.read_csv.html)

[pandas.read_excel](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)

[pandas.read_json](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html)

In [36]:
df_movies = pd.read_csv("Data\\movie.csv")

#could even read the file from a website if you can find a direct link. 
#This also saves you from having to download and store on your computer (good)
#but you need to have internet access each time your run (bad)
#df_movies = pd.read_csv("https://raw.githubusercontent.com/MattDBailey/ANOP330/main/Data/movie.csv")

### What's in this data?

In [37]:
df_movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

### Can we get a cleaner view of this data? Look at the top few rows `.head()` or the last few rows `.tail()` or a random sample `.sample()`

In [27]:
df_movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [28]:
df_movies.tail()

color     director_name  num_critic_for_reviews  duration  \
4911  Color       Scott Smith                     1.0      87.0   
4912  Color               NaN                    43.0      43.0   
4913  Color  Benjamin Roberds                    13.0      76.0   
4914  Color       Daniel Hsia                    14.0     100.0   
4915  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes    gross                        genres  ...  \
4911                   637.0      NaN                  Comedy|Drama  ...   
4912                   841.0      NaN  Crime|Drama|Mystery|Thriller  ...   
4913                     0.0      NaN         Drama|Horror|Thriller  ...   
4914                   946.0  10443.0          Comedy|Drama|Romance  ...   
4915                    86.0  85222.0                   Documentary  ...   

     num_user_for_reviews language  country  content_rating  budget  \
4911                  6.0  English   Canada             NaN     NaN   
4912                359.0  English      USA           TV-14     NaN   
4913                  3.0  English      USA             NaN  1400.0   
4914                  9.0  English      USA           PG-13     NaN   
4915                 84.0  English      USA              PG  1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4911      2013.0                  470.0        7.7           NaN   
4912         NaN                  593.0        7.5         16.00   
4913      2013.0                    0.0        6.3           NaN   
4914      2012.0                  719.0        6.3          2.35   
4915      2004.0                   23.0        6.6          1.85   

     movie_facebook_likes  
4911                   84  
4912                32000  
4913                   16  
4914                  660  
4915                  456  

[5 rows x 28 columns]

In [31]:
df_movies.sample(5)

color       director_name  num_critic_for_reviews  duration  \
1139  Color     Louis Leterrier                   177.0     103.0   
4430  Color       Mona Fastvold                    16.0      91.0   
1366  Color     Nick Cassavetes                   129.0     116.0   
580   Color        Paul Tibbitt                   147.0      92.0   
4542  Color  Mariette Monpierre                     2.0      78.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
1139                    255.0                  5000.0             Jet Li   
4430                      2.0                     9.0       Brady Corbet   
1366                    415.0                   644.0      Laura Harring   
580                       5.0                   726.0         Billy West   
4542                      0.0                     0.0  Christophe Cherki   

      actor_1_facebook_likes        gross  \
1139                 11000.0   24520892.0   
4430                   321.0          NaN   
1366                 18000.0   71026631.0   
580                    870.0  162495848.0   
4542                     0.0          NaN   

                                         genres  ... num_user_for_reviews  \
1139                Action|Crime|Drama|Thriller  ...                303.0   
4430                                      Drama  ...                  2.0   
1366                       Crime|Drama|Thriller  ...                394.0   
580   Adventure|Animation|Comedy|Family|Fantasy  ...                133.0   
4542                                      Drama  ...                  2.0   

     language  country  content_rating      budget  title_year  \
1139  English   France               R  45000000.0      2005.0   
4430  English   Norway       Not Rated   1100000.0      2014.0   
1366  English      USA           PG-13  36000000.0      2002.0   
580   English      USA              PG  74000000.0      2015.0   
4542   French   France             NaN    900000.0      2011.0   

     actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
1139                 5000.0        7.0          2.35                    0  
4430                  287.0        4.9          2.35                  234  
1366                  669.0        7.0          1.85                    0  
580                   861.0        6.0          1.85                16000  
4542                    0.0        6.7         16.00                   24  

[5 rows x 28 columns]

### What are some ATTRIBUTES of the DataFrame?

In [38]:
df_movies.shape

(4916, 28)

In [51]:
#show all the columns and store them
cols = df_movies.columns
cols

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [54]:
#show the index (row names)
ind = df_movies.index
ind

RangeIndex(start=0, stop=4916, step=1)

### See slides for the data dictionary. You should ALWAYS provide your reader a data dictionary (included created fields). The correctness of your analysis is COMPLETELY dependent on the what each data field (and row) represents! It's the data is the product for building, so you better understand your product.

## What are the types of data?
### pandas tries to guess what the data types are, so check that it guessed correctly (dates as datetime, integers as integers, etc.). Watch out for numerical fields that are actually categorical (or objects). Such as 0=Monday, 1=Tuesday, etc.. See slides for types.

In [58]:
df_movies.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

In [69]:
df_movies['content_rating'].value_counts()

R            2067
PG-13        1411
PG            686
Not Rated     115
G             112
Unrated        59
Approved       54
TV-14          30
TV-MA          18
TV-PG          13
X              12
TV-G           10
Passed          9
NC-17           7
GP              6
M               5
TV-Y7           1
TV-Y            1
Name: content_rating, dtype: int64

In [71]:
df_movies['content_rating'].astype('category').cat.categories

Index(['Approved', 'G', 'GP', 'M', 'NC-17', 'Not Rated', 'PG', 'PG-13',
       'Passed', 'R', 'TV-14', 'TV-G', 'TV-MA', 'TV-PG', 'TV-Y', 'TV-Y7',
       'Unrated', 'X'],
      dtype='object')

## Let's get some information about the fields (columns)

In [72]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float64
 3   duration                   4901 non-null   float64
 4   director_facebook_likes    4814 non-null   float64
 5   actor_3_facebook_likes     4893 non-null   float64
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float64
 8   gross                      4054 non-null   float64
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int64  
 13  cast_total_facebook_likes  4916 non-null   int64

## Here is a function that I found to decrease the memory of your data set by picking a smaller data type (e.g., 8 bit float from 64 bit float)
## It isn't much of an issue for our in-class data, but handy for larger data sets.

In [73]:
# Function to decrease data usage in notebook (from Baek Kyun Shin)
def downcast(df, verbose=True):
    start_mem = df.memory_usage().sum() / 1024**2
    for col in df.columns:
        dtype_name = df[col].dtype.name
        if dtype_name == 'object':
            pass
        elif dtype_name == 'bool':
            df[col] = df[col].astype('int8')
        elif dtype_name.startswith('int') or (df[col].round() == df[col]).all():
            df[col] = pd.to_numeric(df[col], downcast='integer')
        else:
            df[col] = pd.to_numeric(df[col], downcast='float')
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose:
        print('{:.1f}% Compressed'.format(100 * (start_mem - end_mem) / start_mem))
    
    return df

In [75]:
df_movies = downcast(df_movies)

28.6% Compressed


In [76]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float32
 3   duration                   4901 non-null   float32
 4   director_facebook_likes    4814 non-null   float32
 5   actor_3_facebook_likes     4893 non-null   float32
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float32
 8   gross                      4054 non-null   float32
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int32  
 13  cast_total_facebook_likes  4916 non-null   int32

##  Reduced from 1.1 MB (1100 KB) to 0.768 MB (768 KB) a 28.6% reduction in memory! Mostly from moving from float64 and int64 to float32 and int32.

In [12]:
director = df_movies.director_name

In [13]:
director.head()

0        James Cameron
1       Gore Verbinski
2           Sam Mendes
3    Christopher Nolan
4          Doug Walker
Name: director_name, dtype: object

In [14]:
director.dtypes

dtype('O')

In [15]:
director.size

4916

In [16]:
fb_A1likes = df_movies.actor_1_facebook_likes

In [17]:
fb_A1likes.count()

4909

In [18]:
fb_A1likes_drop = fb_A1likes.dropna()
fb_A1likes_drop.size

4909

In [19]:
fb_A1likes_z = (fb_A1likes_drop - fb_A1likes_drop.mean())/fb_A1likes_drop.std()

In [20]:
fb_A1likes_z

0      -0.363705
1       2.217882
2       0.298240
3       1.357353
4      -0.421228
          ...   
4911   -0.387734
4912   -0.374230
4913   -0.429900
4914   -0.367280
4915   -0.424207
Name: actor_1_facebook_likes, Length: 4909, dtype: float64

In [21]:
fb_A1likes_z.gt(3).sum() + fb_A1likes_z.lt(-3).sum()

17

In [22]:
fb_A1likes_z.lt(-3).sum()

0

In [23]:
fb_A1likes.fillna(0).astype(int).head()

0     1000
1    40000
2    11000
3    27000
4      131
Name: actor_1_facebook_likes, dtype: int32

In [24]:
(fb_A1likes.fillna(0)
     .astype(int)
     .head()
)

0     1000
1    40000
2    11000
3    27000
4      131
Name: actor_1_facebook_likes, dtype: int32

In [25]:
col_map = {
    "director_name": "director",
    "director_facebook_likes": "director_FB_likes",
}


In [26]:
df_movies.rename(columns=col_map).head(10)

color           director  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   
5  Color     Andrew Stanton                   462.0     132.0   
6  Color          Sam Raimi                   392.0     156.0   
7  Color       Nathan Greno                   324.0     100.0   
8  Color        Joss Whedon                   635.0     141.0   
9  Color        David Yates                   375.0     153.0   

   director_FB_likes  actor_3_facebook_likes       actor_2_name  \
0                0.0                   855.0   Joel David Moore   
1              563.0                  1000.0      Orlando Bloom   
2                0.0                   161.0       Rory Kinnear   
3            22000.0                 23000.0     Christian Bale   
4              131.0                     NaN         Rob Walker   
5              475.0                   530.0    Samantha Morton   
6                0.0                  4000.0       James Franco   
7               15.0                   284.0       Donna Murphy   
8                0.0                 19000.0  Robert Downey Jr.   
9              282.0                 10000.0   Daniel Radcliffe   

   actor_1_facebook_likes        gross  \
0                  1000.0  760505847.0   
1                 40000.0  309404152.0   
2                 11000.0  200074175.0   
3                 27000.0  448130642.0   
4                   131.0          NaN   
5                   640.0   73058679.0   
6                 24000.0  336530303.0   
7                   799.0  200807262.0   
8                 26000.0  458991599.0   
9                 25000.0  301956980.0   

                                              genres  ...  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...   
1                           Action|Adventure|Fantasy  ...   
2                          Action|Adventure|Thriller  ...   
3                                    Action|Thriller  ...   
4                                        Documentary  ...   
5                            Action|Adventure|Sci-Fi  ...   
6                           Action|Adventure|Romance  ...   
7  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
8                            Action|Adventure|Sci-Fi  ...   
9                   Adventure|Family|Fantasy|Mystery  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   
5                738.0  English      USA           PG-13  263700000.0   
6               1902.0  English      USA           PG-13  258000000.0   
7                387.0  English      USA              PG  260000000.0   
8               1117.0  English      USA           PG-13  250000000.0   
9                973.0  English       UK              PG  250000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   
5      2012.0                  632.0        6.6          2.35   
6      2007.0                11000.0        6.2          2.35   
7      2010.0                  553.0        7.8          1.85   
8      2015.0            

In [27]:
total = (
    df_movies.actor_1_facebook_likes
    + df_movies.actor_2_facebook_likes
    + df_movies.actor_3_facebook_likes
    + df_movies.director_facebook_likes
)

In [28]:
total

0        2791.0
1       46563.0
2       11554.0
3       95000.0
4           NaN
         ...   
4911     1427.0
4912        NaN
4913        0.0
4914     2154.0
4915      141.0
Length: 4916, dtype: float64

In [29]:
total.head()

0     2791.0
1    46563.0
2    11554.0
3    95000.0
4        NaN
dtype: float64

In [30]:
df_movies.assign(total_likes=total).head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  language country  content_rating       budget title_year  \
0  English     USA           PG-13  237000000.0     2009.0   
1  English     USA           PG-13  300000000.0     2007.0   
2  English      UK           PG-13  245000000.0     2015.0   
3  English     USA           PG-13  250000000.0     2012.0   
4      NaN     NaN             NaN          NaN        NaN   

   actor_2_facebook_likes imdb_score aspect_ratio  movie_facebook_likes  \
0                   936.0        7.9         1.78                 33000   
1                  5000.0        7.1         2.35                     0   
2                   393.0        6.8         2.35                 85000   
3                 23000.0        8.5         2.35                164000   
4                    12.0        7.1          NaN                     0   

  total_likes  
0      2791.0  
1     46563.0  
2     11554.0  
3     95000.0  
4         NaN  

[5 rows x 29 columns]

In [31]:
df_movies2 = df_movies.assign(total_likes=total)
df_movies2.head()


color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  language country  content_rating       budget title_year  \
0  English     USA           PG-13  237000000.0     2009.0   
1  English     USA           PG-13  300000000.0     2007.0   
2  English      UK           PG-13  245000000.0     2015.0   
3  English     USA           PG-13  250000000.0     2012.0   
4      NaN     NaN             NaN          NaN        NaN   

   actor_2_facebook_likes imdb_score aspect_ratio  movie_facebook_likes  \
0                   936.0        7.9         1.78                 33000   
1                  5000.0        7.1         2.35                     0   
2                   393.0        6.8         2.35                 85000   
3                 23000.0        8.5         2.35                164000   
4                    12.0        7.1          NaN                     0   

  total_likes  
0      2791.0  
1     46563.0  
2     11554.0  
3     95000.0  
4         NaN  

[5 rows x 29 columns]

In [32]:
df_movies2 = df_movies2.drop(columns=["total_likes"]) 

In [33]:
df_movies2.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [34]:
fb_A1likes_drop + 10

0        1010.0
1       40010.0
2       11010.0
3       27010.0
4         141.0
         ...   
4911      647.0
4912      851.0
4913       10.0
4914      956.0
4915       96.0
Name: actor_1_facebook_likes, Length: 4909, dtype: float64

In [35]:
cols = [
    "actor_1_name",
    "actor_2_name",
    "actor_3_name",
    "director_name"
]
cols

['actor_1_name', 'actor_2_name', 'actor_3_name', 'director_name']

In [36]:
df_movie_names = df_movies[cols]

In [37]:
df_movie_names.head()

actor_1_name      actor_2_name          actor_3_name      director_name
0      CCH Pounder  Joel David Moore             Wes Studi      James Cameron
1      Johnny Depp     Orlando Bloom        Jack Davenport     Gore Verbinski
2  Christoph Waltz      Rory Kinnear      Stephanie Sigman         Sam Mendes
3        Tom Hardy    Christian Bale  Joseph Gordon-Levitt  Christopher Nolan
4      Doug Walker        Rob Walker                   NaN        Doug Walker

In [38]:
df_movies.loc[:,cols]

actor_1_name      actor_2_name          actor_3_name  \
0         CCH Pounder  Joel David Moore             Wes Studi   
1         Johnny Depp     Orlando Bloom        Jack Davenport   
2     Christoph Waltz      Rory Kinnear      Stephanie Sigman   
3           Tom Hardy    Christian Bale  Joseph Gordon-Levitt   
4         Doug Walker        Rob Walker                   NaN   
...               ...               ...                   ...   
4911      Eric Mabius     Daphne Zuniga          Crystal Lowe   
4912      Natalie Zea     Valorie Curry         Sam Underwood   
4913      Eva Boehnke     Maxwell Moody        David Chandler   
4914        Alan Ruck     Daniel Henney           Eliza Coupe   
4915      John August  Brian Herzlinger              Jon Gunn   

          director_name  
0         James Cameron  
1        Gore Verbinski  
2            Sam Mendes  
3     Christopher Nolan  
4           Doug Walker  
...                 ...  
4911        Scott Smith  
4912                NaN  
4913   Benjamin Roberds  
4914        Daniel Hsia  
4915           Jon Gunn  

[4916 rows x 4 columns]

In [39]:
df_movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [40]:
df_movies.iloc[0,2]

723.0

In [41]:
df_movies.dtypes.value_counts()

float64    13
object     12
int64       3
dtype: int64

In [42]:
df_movies.select_dtypes(include=["object"]).head()

color      director_name      actor_2_name  \
0  Color      James Cameron  Joel David Moore   
1  Color     Gore Verbinski     Orlando Bloom   
2  Color         Sam Mendes      Rory Kinnear   
3  Color  Christopher Nolan    Christian Bale   
4    NaN        Doug Walker        Rob Walker   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4                      Documentary      Doug Walker   

                                  movie_title          actor_3_name  \
0                                      Avatar             Wes Studi   
1    Pirates of the Caribbean: At World's End        Jack Davenport   
2                                     Spectre      Stephanie Sigman   
3                       The Dark Knight Rises  Joseph Gordon-Levitt   
4  Star Wars: Episode VII - The Force Awakens                   NaN   

                                       plot_keywords  \
0             avatar|future|marine|native|paraplegic   
1  goddess|marriage ceremony|marriage proposal|pi...   
2                bomb|espionage|sequel|spy|terrorist   
3  deception|imprisonment|lawlessness|police offi...   
4                                                NaN   

                                     movie_imdb_link language country  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...  English     USA   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...  English     USA   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...  English      UK   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...  English     USA   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...      NaN     NaN   

  content_rating  
0          PG-13  
1          PG-13  
2          PG-13  
3          PG-13  
4            NaN

In [43]:
df_movies.describe().T

count          mean           std      min  \
num_critic_for_reviews     4867.0  1.379889e+02  1.202394e+02     1.00   
duration                   4901.0  1.070908e+02  2.528602e+01     7.00   
director_facebook_likes    4814.0  6.910145e+02  2.832954e+03     0.00   
actor_3_facebook_likes     4893.0  6.312763e+02  1.625875e+03     0.00   
actor_1_facebook_likes     4909.0  6.494488e+03  1.510699e+04     0.00   
gross                      4054.0  4.764451e+07  6.737255e+07   162.00   
num_voted_users            4916.0  8.264492e+04  1.383222e+05     5.00   
cast_total_facebook_likes  4916.0  9.579816e+03  1.816432e+04     0.00   
facenumber_in_poster       4903.0  1.377320e+00  2.023826e+00     0.00   
num_user_for_reviews       4895.0  2.676688e+02  3.729348e+02     1.00   
budget                     4432.0  3.654749e+07  1.002427e+08   218.00   
title_year                 4810.0  2.002448e+03  1.245398e+01  1916.00   
actor_2_facebook_likes     4903.0  1.621924e+03  4.011300e+03     0.00   
imdb_score                 4916.0  6.437429e+00  1.127802e+00     1.60   
aspect_ratio               4590.0  2.222349e+00  1.402940e+00     1.18   
movie_facebook_likes       4916.0  7.348294e+03  1.920602e+04     0.00   

                                  25%          50%          75%           max  
num_critic_for_reviews          49.00       108.00       191.00  8.130000e+02  
duration                        93.00       103.00       118.00  5.110000e+02  
director_facebook_likes          7.00        48.00       189.75  2.300000e+04  
actor_3_facebook_likes         132.00       366.00       633.00  2.300000e+04  
actor_1_facebook_likes         607.00       982.00     11000.00  6.400000e+05  
gross                      5019656.25  25043962.00  61108412.75  7.605058e+08  
num_voted_users               8361.75     33132.50     93772.75  1.689764e+06  
cast_total_facebook_likes     1394.75      3049.00     13616.75  6.567300e+05  
facenumber_in_poster             0.00         1.00         2.00  4.300000e+01  
num_user_for_reviews            64.00       153.00       320.50  5.060000e+03  
budget                     6000000.00  19850000.00  43000000.00  4.200000e+09  
title_year                    1999.00      2005.00      2011.00  2.016000e+03  
actor_2_facebook_likes         277.00       593.00       912.00  1.370000e+05  
imdb_score                       5.80         6.60         7.20  9.500000e+00  
aspect_ratio                     1.85         2.35         2.35  1.600000e+01  
movie_facebook_likes             0.00       159.00      2000.00  3.490000e+05

In [44]:
# Check the version of pandas that you are using
pd.__version__

'1.0.5'

In [45]:
df_movies.describe().T

count          mean           std      min  \
num_critic_for_reviews     4867.0  1.379889e+02  1.202394e+02     1.00   
duration                   4901.0  1.070908e+02  2.528602e+01     7.00   
director_facebook_likes    4814.0  6.910145e+02  2.832954e+03     0.00   
actor_3_facebook_likes     4893.0  6.312763e+02  1.625875e+03     0.00   
actor_1_facebook_likes     4909.0  6.494488e+03  1.510699e+04     0.00   
gross                      4054.0  4.764451e+07  6.737255e+07   162.00   
num_voted_users            4916.0  8.264492e+04  1.383222e+05     5.00   
cast_total_facebook_likes  4916.0  9.579816e+03  1.816432e+04     0.00   
facenumber_in_poster       4903.0  1.377320e+00  2.023826e+00     0.00   
num_user_for_reviews       4895.0  2.676688e+02  3.729348e+02     1.00   
budget                     4432.0  3.654749e+07  1.002427e+08   218.00   
title_year                 4810.0  2.002448e+03  1.245398e+01  1916.00   
actor_2_facebook_likes     4903.0  1.621924e+03  4.011300e+03     0.00   
imdb_score                 4916.0  6.437429e+00  1.127802e+00     1.60   
aspect_ratio               4590.0  2.222349e+00  1.402940e+00     1.18   
movie_facebook_likes       4916.0  7.348294e+03  1.920602e+04     0.00   

                                  25%          50%          75%           max  
num_critic_for_reviews          49.00       108.00       191.00  8.130000e+02  
duration                        93.00       103.00       118.00  5.110000e+02  
director_facebook_likes          7.00        48.00       189.75  2.300000e+04  
actor_3_facebook_likes         132.00       366.00       633.00  2.300000e+04  
actor_1_facebook_likes         607.00       982.00     11000.00  6.400000e+05  
gross                      5019656.25  25043962.00  61108412.75  7.605058e+08  
num_voted_users               8361.75     33132.50     93772.75  1.689764e+06  
cast_total_facebook_likes     1394.75      3049.00     13616.75  6.567300e+05  
facenumber_in_poster             0.00         1.00         2.00  4.300000e+01  
num_user_for_reviews            64.00       153.00       320.50  5.060000e+03  
budget                     6000000.00  19850000.00  43000000.00  4.200000e+09  
title_year                    1999.00      2005.00      2011.00  2.016000e+03  
actor_2_facebook_likes         277.00       593.00       912.00  1.370000e+05  
imdb_score                       5.80         6.60         7.20  9.500000e+00  
aspect_ratio                     1.85         2.35         2.35  1.600000e+01  
movie_facebook_likes             0.00       159.00      2000.00  3.490000e+05

In [46]:
df_movies.country.fillna("").max()

'West Germany'

In [47]:
df_college = pd.read_csv("Data\\college.csv",index_col="INSTNM")

In [48]:
df_college.columns

Index(['CITY', 'STABBR', 'HBCU', 'MENONLY', 'WOMENONLY', 'RELAFFIL',
       'SATVRMID', 'SATMTMID', 'DISTANCEONLY', 'UGDS', 'UGDS_WHITE',
       'UGDS_BLACK', 'UGDS_HISP', 'UGDS_ASIAN', 'UGDS_AIAN', 'UGDS_NHPI',
       'UGDS_2MOR', 'UGDS_NRA', 'UGDS_UNKN', 'PPTUG_EF', 'CURROPER', 'PCTPELL',
       'PCTFLOAN', 'UG25ABV', 'MD_EARN_WNE_P10', 'GRAD_DEBT_MDN_SUPP'],
      dtype='object')

In [49]:
df_diversity = df_college.filter(like="UGDS_")

In [50]:
df_diversity.head()

UGDS_WHITE  UGDS_BLACK  UGDS_HISP  \
INSTNM                                                                   
Alabama A & M University                 0.0333      0.9353     0.0055   
University of Alabama at Birmingham      0.5922      0.2600     0.0283   
Amridge University                       0.2990      0.4192     0.0069   
University of Alabama in Huntsville      0.6988      0.1255     0.0382   
Alabama State University                 0.0158      0.9208     0.0121   

                                     UGDS_ASIAN  UGDS_AIAN  UGDS_NHPI  \
INSTNM                                                                  
Alabama A & M University                 0.0019     0.0024     0.0019   
University of Alabama at Birmingham      0.0518     0.0022     0.0007   
Amridge University                       0.0034     0.0000     0.0000   
University of Alabama in Huntsville      0.0376     0.0143     0.0002   
Alabama State University                 0.0019     0.0010     0.0006   

                                     UGDS_2MOR  UGDS_NRA  UGDS_UNKN  
INSTNM                                                               
Alabama A & M University                0.0000    0.0059     0.0138  
University of Alabama at Birmingham     0.0368    0.0179     0.0100  
Amridge University                      0.0000    0.0000     0.2715  
University of Alabama in Huntsville     0.0172    0.0332     0.0350  
Alabama State University                0.0098    0.0243     0.0137

In [51]:
df_diversity.isna().sum(axis="columns").eq(9).value_counts()

False    6874
True      661
dtype: int64

In [52]:
(
	df_diversity.isna()
	.sum(axis="columns")
	.sort_values(ascending=False)
	.head()
)

INSTNM
Excel Learning Center-San Antonio South         9
Philadelphia College of Osteopathic Medicine    9
Assemblies of God Theological Seminary          9
Episcopal Divinity School                       9
Phillips Graduate Institute                     9
dtype: int64

In [53]:
df_diversity = df_diversity.dropna(how="all")

In [54]:
df_diversity.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6874 entries, Alabama A & M University to Coastal Pines Technical College
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   UGDS_WHITE  6874 non-null   float64
 1   UGDS_BLACK  6874 non-null   float64
 2   UGDS_HISP   6874 non-null   float64
 3   UGDS_ASIAN  6874 non-null   float64
 4   UGDS_AIAN   6874 non-null   float64
 5   UGDS_NHPI   6874 non-null   float64
 6   UGDS_2MOR   6874 non-null   float64
 7   UGDS_NRA    6874 non-null   float64
 8   UGDS_UNKN   6874 non-null   float64
dtypes: float64(9)
memory usage: 537.0+ KB


In [55]:
metric = (
    df_diversity.ge(0.15)
    .sum(axis="columns")
)

In [56]:
metric.sort_values(ascending=False).head()

INSTNM
Regency Beauty Institute-Austin          5
Central Texas Beauty College-Temple      5
Sullivan and Cogliano Training Center    4
Ambria College of Nursing                4
Berkeley College-New York                4
dtype: int64

In [57]:
df_diversity.loc[["Regency Beauty Institute-Austin","Central Texas Beauty College-Temple"]]

UGDS_WHITE  UGDS_BLACK  UGDS_HISP  \
INSTNM                                                                   
Regency Beauty Institute-Austin          0.1867      0.2133     0.1600   
Central Texas Beauty College-Temple      0.1616      0.2323     0.2626   

                                     UGDS_ASIAN  UGDS_AIAN  UGDS_NHPI  \
INSTNM                                                                  
Regency Beauty Institute-Austin          0.0000        0.0        0.0   
Central Texas Beauty College-Temple      0.0202        0.0        0.0   

                                     UGDS_2MOR  UGDS_NRA  UGDS_UNKN  
INSTNM                                                               
Regency Beauty Institute-Austin         0.1733       0.0     0.2667  
Central Texas Beauty College-Temple     0.1717       0.0     0.1515

In [58]:
df_diversity.loc[["Bucknell University"]]

UGDS_WHITE  UGDS_BLACK  UGDS_HISP  UGDS_ASIAN  UGDS_AIAN  \
INSTNM                                                                          
Bucknell University      0.7885      0.0317     0.0527      0.0357     0.0003   

                     UGDS_NHPI  UGDS_2MOR  UGDS_NRA  UGDS_UNKN  
INSTNM                                                          
Bucknell University        0.0     0.0343     0.053     0.0037

In [59]:
df_college.describe(include="number").T

count         mean          std    min         25%        50%  \
HBCU          7164.0     0.014238     0.118478    0.0    0.000000    0.00000   
MENONLY       7164.0     0.009213     0.095546    0.0    0.000000    0.00000   
WOMENONLY     7164.0     0.005304     0.072642    0.0    0.000000    0.00000   
RELAFFIL      7535.0     0.190975     0.393096    0.0    0.000000    0.00000   
SATVRMID      1185.0   522.819409    68.578862  290.0  475.000000  510.00000   
SATMTMID      1196.0   530.765050    73.469767  310.0  482.000000  520.00000   
DISTANCEONLY  7164.0     0.005583     0.074519    0.0    0.000000    0.00000   
UGDS          6874.0  2356.837940  5474.275871    0.0  117.000000  412.50000   
UGDS_WHITE    6874.0     0.510207     0.286958    0.0    0.267500    0.55570   
UGDS_BLACK    6874.0     0.189997     0.224587    0.0    0.036125    0.10005   
UGDS_HISP     6874.0     0.161635     0.221854    0.0    0.027600    0.07140   
UGDS_ASIAN    6874.0     0.033544     0.073777    0.0    0.002500    0.01290   
UGDS_AIAN     6874.0     0.013813     0.070196    0.0    0.000000    0.00260   
UGDS_NHPI     6874.0     0.004569     0.033125    0.0    0.000000    0.00000   
UGDS_2MOR     6874.0     0.023950     0.031288    0.0    0.000000    0.01750   
UGDS_NRA      6874.0     0.016086     0.050172    0.0    0.000000    0.00000   
UGDS_UNKN     6874.0     0.045181     0.093440    0.0    0.000000    0.01430   
PPTUG_EF      6853.0     0.226639     0.246470    0.0    0.000000    0.15040   
CURROPER      7535.0     0.923291     0.266146    0.0    1.000000    1.00000   
PCTPELL       6849.0     0.530643     0.225544    0.0    0.357800    0.52150   
PCTFLOAN      6849.0     0.522211     0.283616    0.0    0.332900    0.58330   
UG25ABV       6718.0     0.410021     0.228939    0.0    0.241500    0.40075   

                      75%          max  
HBCU             0.000000       1.0000  
MENONLY          0.000000       1.0000  
WOMENONLY        0.000000       1.0000  
RELAFFIL         0.000000       1.0000  
SATVRMID       555.000000     765.0000  
SATMTMID       565.000000     785.0000  
DISTANCEONLY     0.000000       1.0000  
UGDS          1929.500000  151558.0000  
UGDS_WHITE       0.747875       1.0000  
UGDS_BLACK       0.257700       1.0000  
UGDS_HISP        0.198875       1.0000  
UGDS_ASIAN       0.032700       0.9727  
UGDS_AIAN        0.007300       1.0000  
UGDS_NHPI        0.002500       0.9983  
UGDS_2MOR        0.033900       0.5333  
UGDS_NRA         0.011700       0.9286  
UGDS_UNKN        0.045400       0.9027  
PPTUG_EF         0.376900       1.0000  
CURROPER         1.000000       1.0000  
PCTPELL          0.712900       1.0000  
PCTFLOAN         0.745000       1.0000  
UG25ABV          0.572275       1.0000

In [60]:
df_college.describe(include="object").T

count unique                top  freq
CITY                7535   2514           New York    87
STABBR              7535     59                 CA   773
MD_EARN_WNE_P10     6413    598  PrivacySuppressed   822
GRAD_DEBT_MDN_SUPP  7503   2038  PrivacySuppressed  1510

In [61]:
object_cols = df_college.select_dtypes('object').columns.tolist()

In [62]:
df_college[object_cols] = df_college[object_cols].astype('category')

In [63]:
df_college.dtypes

CITY                  category
STABBR                category
HBCU                   float64
MENONLY                float64
WOMENONLY              float64
RELAFFIL                 int64
SATVRMID               float64
SATMTMID               float64
DISTANCEONLY           float64
UGDS                   float64
UGDS_WHITE             float64
UGDS_BLACK             float64
UGDS_HISP              float64
UGDS_ASIAN             float64
UGDS_AIAN              float64
UGDS_NHPI              float64
UGDS_2MOR              float64
UGDS_NRA               float64
UGDS_UNKN              float64
PPTUG_EF               float64
CURROPER                 int64
PCTPELL                float64
PCTFLOAN               float64
UG25ABV                float64
MD_EARN_WNE_P10       category
GRAD_DEBT_MDN_SUPP    category
dtype: object

In [64]:
obj_cols =(
df_movies.select_dtypes(include=["object"])
.columns
.tolist()
)

In [65]:
obj_cols

['color',
 'director_name',
 'actor_2_name',
 'genres',
 'actor_1_name',
 'movie_title',
 'actor_3_name',
 'plot_keywords',
 'movie_imdb_link',
 'language',
 'country',
 'content_rating']

In [66]:
df_movies[obj_cols] = df_movies[obj_cols].astype('category')

In [67]:
df_movies.color.cat.categories

Index(['Black and White', 'Color'], dtype='object')

In [68]:
df_movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

In [69]:
pd.get_dummies(df_movies.color)

Black and White  Color
0                   0      1
1                   0      1
2                   0      1
3                   0      1
4                   0      0
...               ...    ...
4911                0      1
4912                0      1
4913                0      1
4914                0      1
4915                0      1

[4916 rows x 2 columns]

In [70]:
df_movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [71]:
df_movies.director_facebook_likes.max()

23000.0

In [72]:
df_movies[
    (df_movies.color=='Color') & (df_movies.duration>=180)
    ].head()

color  director_name  num_critic_for_reviews  duration  \
10  Color    Zack Snyder                   673.0     183.0   
23  Color  Peter Jackson                   509.0     186.0   
25  Color  Peter Jackson                   446.0     201.0   
26  Color  James Cameron                   315.0     194.0   
32  Color    Shane Black                   608.0     195.0   

    director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
10                      0.0                  2000.0        Lauren Cohan   
23                      0.0                   773.0          Adam Brown   
25                      0.0                    84.0  Thomas Kretschmann   
26                      0.0                   794.0        Kate Winslet   
32                   1000.0                  3000.0         Jon Favreau   

    actor_1_facebook_likes        gross                          genres  ...  \
10                 15000.0  330249062.0         Action|Adventure|Sci-Fi  ...   
23                  5000.0  258355354.0               Adventure|Fantasy  ...   
25                  6000.0  218051260.0  Action|Adventure|Drama|Romance  ...   
26                 29000.0  658672302.0                   Drama|Romance  ...   
32                 21000.0  408992272.0         Action|Adventure|Sci-Fi  ...   

   num_user_for_reviews language      country  content_rating       budget  \
10               3018.0  English          USA           PG-13  250000000.0   
23                951.0  English          USA           PG-13  225000000.0   
25               2618.0  English  New Zealand           PG-13  207000000.0   
26               2528.0  English          USA           PG-13  200000000.0   
32               1187.0  English          USA           PG-13  200000000.0   

    title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
10      2016.0                 4000.0        6.9          2.35   
23      2013.0                  972.0        7.9          2.35   
25      2005.0                  919.0        7.2          2.35   
26      1997.0                14000.0        7.7          2.35   
32      2013.0                 4000.0        7.2          2.35   

   movie_facebook_likes  
10               197000  
23                83000  
25                    0  
26                26000  
32                95000  

[5 rows x 28 columns]

In [73]:
df_movies.loc[
    (df_movies.color=='Color') & (df_movies.duration>=180)
    ].head()

color  director_name  num_critic_for_reviews  duration  \
10  Color    Zack Snyder                   673.0     183.0   
23  Color  Peter Jackson                   509.0     186.0   
25  Color  Peter Jackson                   446.0     201.0   
26  Color  James Cameron                   315.0     194.0   
32  Color    Shane Black                   608.0     195.0   

    director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
10                      0.0                  2000.0        Lauren Cohan   
23                      0.0                   773.0          Adam Brown   
25                      0.0                    84.0  Thomas Kretschmann   
26                      0.0                   794.0        Kate Winslet   
32                   1000.0                  3000.0         Jon Favreau   

    actor_1_facebook_likes        gross                          genres  ...  \
10                 15000.0  330249062.0         Action|Adventure|Sci-Fi  ...   
23                  5000.0  258355354.0               Adventure|Fantasy  ...   
25                  6000.0  218051260.0  Action|Adventure|Drama|Romance  ...   
26                 29000.0  658672302.0                   Drama|Romance  ...   
32                 21000.0  408992272.0         Action|Adventure|Sci-Fi  ...   

   num_user_for_reviews language      country  content_rating       budget  \
10               3018.0  English          USA           PG-13  250000000.0   
23                951.0  English          USA           PG-13  225000000.0   
25               2618.0  English  New Zealand           PG-13  207000000.0   
26               2528.0  English          USA           PG-13  200000000.0   
32               1187.0  English          USA           PG-13  200000000.0   

    title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
10      2016.0                 4000.0        6.9          2.35   
23      2013.0                  972.0        7.9          2.35   
25      2005.0                  919.0        7.2          2.35   
26      1997.0                14000.0        7.7          2.35   
32      2013.0                 4000.0        7.2          2.35   

   movie_facebook_likes  
10               197000  
23                83000  
25                    0  
26                26000  
32                95000  

[5 rows x 28 columns]

In [74]:
df_movies.loc[df_movies['color'] == 'Color']

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
5     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
5                       475.0                   530.0   Samantha Morton   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
5                      640.0   73058679.0          Action|Adventure|Sci-Fi   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
5     ...                738.0  English      USA           PG-13  263700000.0   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
5         2012.0                  632.

In [75]:
df_titanic.columns

NameError: name 'df_titanic' is not defined

In [86]:
df = pd.DataFrame({'a':[1, 2, 3], 'b':[1, 4, 3], 'c':[1, 4, 3]})
df

a  b  c
0  1  1  1
1  2  4  4
2  3  3  3

In [87]:
from sklearn.preprocessing import Normalizer

In [92]:
trans = Normalizer(norm='l2')
pd.DataFrame(trans.fit_transform(df),columns=df.columns,index=df.index)

a         b         c
0  0.577350  0.577350  0.577350
1  0.333333  0.666667  0.666667
2  0.577350  0.577350  0.577350

In [110]:
max = df.max(axis=1)
df.divide(max,axis=0)

0    1
1    4
2    3
dtype: int64

a    b    c
0  1.0  1.0  1.0
1  0.5  1.0  1.0
2  1.0  1.0  1.0